docs/internal/DistributedArchitectureGuide.md

```diff
@@ -12,6 +12,214 @@ A guide to the general Elasticsearch components can be found [here](https://gith
+snapshotting to remote repositories. These repositories usually come with their
+own SDK and networking stack. For example AWS SDK comes with tomcat or netty,
+Azure with netty-based project-reactor, GCP uses default java HTTP
+client. Depending on the SDK underlying client might be reused between
+repositories. Some provide better control over networking settings, others
+less.
+
+## Sync/Async IO and threading
+
+Elasticsearch does a lot of IO - disk, HTTP server, Transport client/server,
+repositories. Unfortunately not everything fits into synchronous or asynchronous
+style. As a result we have a bag of everything.
+
+HTTP and Transport use asynchronous IO. Disk and repositories synchronous IO.
+
+Asynchronous IO utilize a small set of threads by running small tasks,
+minimizing context switch. Synchronous IO uses many threads and let scheduler
+decide what to run. As a result ES runs at 100+ threads where Async threads
+co-exist(compete) with Sync threads.
+
+## Netty
+
+Netty is a networking framework/toolkit. It's a collection of building blocks
+that provides great control and foot-shooters to build networking applications.
+
+We use netty extensively in our HTTP and Transport networks. It's not a guide
+for netty, but covers major interaction with rest of ES.
+
+### Event-Loop or Transport-Thread
+
+Netty is an Async IO framework. So it runs with a few threads. An event-loop is
+a thread that processes events for one or many connections. Every connection has
+exactly one event-loop that never changes. With this guaranty there is no need
+to synchronize events that happen within the connection/pipeline.
+
+To take full advantage of async IO we have a single Transport ThreadPool that
+serves HTTP and Transport servers and clients. The pool size is CPU bound, for
+example 4 cores would have 4 threads. These 4 threads would serve hundreds or
+thousands of connections.
+
+Event-loop threads serve many connections each, so it's absolutely critical to
+not block thread for a long time. Any blocking operation or heavy compute has to
+be forked to another thread pool. But forking comes with overhead, doing forking
+on every tiny request is a wasted CPU work. As a rule of thumb: don't fork
+simple requests that can be served from memory and do not require heavy
+computations (seconds), otherwise fork.
+
+Transport threads are monitored by `ThreadWatchdog`. When thread runs single
+task longer than 5 seconds warning logs appears. There are many reasons for
+slowness, not only lack of forking. For example GC pauses, CPU starvation from
+other thread-pools.
+
+### ByteBuf - byte buffers and reference counting
+
+One of the performance edges of netty is controlled memory allocation. Netty
+manages byte buffer pools and reuse them heavily. This performance gain comes
+with a cost. And cost is reference counting on developer's shoulders. Netty
+reads socket bytes into pooled byte-buffers and passes them to application. Then
```

Comment:
Might be worth saying a bit more about how this looks in a heap dump - there's a pool of 1MiB `byte[]` objects which Netty slices into 16kiB pages, not all of which may be in use, so you have to account for this when investigating memory usage.